### PR TITLE
[Merged by Bors] - chore(Order/SuccPred/Basic): Move succ/pred being archimedean to a different file

### PR DESCRIPTION
**Pull request by Command-Master**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/SuccPred/Archimedean.lean [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+  exact Iterate.rec (fun b => p a ↔ p b) (fun c hc => hc.trans (hsucc _)) Iff.rfl n
+
+lemma le_total_of_ge {r v₁ v₂ : α} (h₁ : r ≤ v₁) (h₂ : r ≤ v₂) : v₁ ≤ v₂ ∨ v₂ ≤ v₁ := by
+  obtain ⟨n, rfl⟩ := h₁.exists_succ_iterate
+  obtain ⟨m, rfl⟩ := h₂.exists_succ_iterate
+  clear h₁ h₂
+  wlog h : n ≤ m
+  · rw [Or.comm]
+    apply this
+    omega
```
Comment (vihdzp) [inline Mathlib/Order/SuccPred/Archimedean.lean:78]: What's the goal state here? Nothing in this theorem seems to suggest any complex calculations are necessary.

File: Mathlib/Order/SuccPred/Archimedean.lean [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  · exact h0
+  · rw [Function.iterate_succ_apply']
+    exact h1 _ (id_le_iterate_of_id_le le_succ n m) ih
+
+theorem Succ.rec_iff {p : α → Prop} (hsucc : ∀ a, p a ↔ p (succ a)) {a b : α} (h : a ≤ b) :
+    p a ↔ p b := by
+  obtain ⟨n, rfl⟩ := h.exists_succ_iterate
+  exact Iterate.rec (fun b => p a ↔ p b) (fun c hc => hc.trans (hsucc _)) Iff.rfl n
+
+lemma le_total_of_ge {r v₁ v₂ : α} (h₁ : r ≤ v₁) (h₂ : r ≤ v₂) : v₁ ≤ v₂ ∨ v₂ ≤ v₁ := by
```
Comment (vihdzp) [inline Mathlib/Order/SuccPred/Archimedean.lean:71]: So, every directed archimedean order is linear, correct? Can we make that an instance?

**Review by vihdzp — APPROVED** [review]
Thanks! I always considered the archimedean stuff to be a bit out of place in that file. Just some small questions.

File: Mathlib/Order/SuccPred/Archimedean.lean (hunk 1 above)
Comment (Command-Master) [reply Mathlib/Order/SuccPred/Archimedean.lean:78]: I'll replace this with `exact Nat.le_of_not_ge h`.

File: Mathlib/Order/SuccPred/Archimedean.lean (hunk 2 above)
Comment (Command-Master) [reply Mathlib/Order/SuccPred/Archimedean.lean:71]: Couldn't that cause a diamond due to the `decidableLE, decidableEq, decidableLT` in `LinearOrder`?

Comment (vihdzp) [reply Mathlib/Order/SuccPred/Archimedean.lean:71]: Hm, that's true. So can you make that a constructor instead?

Comment (Command-Master) [reply Mathlib/Order/SuccPred/Archimedean.lean:71]: Done.

File: Mathlib/Order/SuccPred/Archimedean.lean [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+  · apply lt_of_le_of_ne h (ne_of_not_le nh).symm
+  · contradiction
+
+/--
+This isn't an instance due to a diamond with the `decidable` instances in `LinearOrder`.
+-/
+noncomputable def IsPredArchimedean.linearOrder [PredOrder α] [IsPredArchimedean α]
+    [IsDirected α (· ≤ ·)] : LinearOrder α :=
+      letI : LinearOrder αᵒᵈ := IsSuccArchimedean.linearOrder
+      inferInstanceAs (LinearOrder αᵒᵈᵒᵈ)
```
Comment (YaelDillies) [inline Mathlib/Order/SuccPred/Archimedean.lean:152]:
```suggestion
/--
This isn't an instance due to a loop with `LinearOrder`.
-/
-- See note [reducible non instances]
noncomputable abbrev IsPredArchimedean.toLinearOrder [PredOrder α] [IsPredArchimedean α]
    [DecidableEq α] [@DecidableRel α (· ≤ ·)] [@DecidableRel α (· < ·)] [IsDirected α (· ≤ ·)] :
    LinearOrder α :=
  letI : LinearOrder αᵒᵈ := IsSuccArchimedean.linearOrder
  inferInstanceAs (LinearOrder αᵒᵈᵒᵈ)
```
and I'll let you fix `IsSuccArchimedean.linearOrder` similarly

**Review by YaelDillies — COMMENTED** [review]
As always, it's much preferrable to do moves/splits separately to additions/modifications.

File: Mathlib/Order/SuccPred/Archimedean.lean [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+-- See note [reducible non instances]
+abbrev IsSuccArchimedean.linearOrder [SuccOrder α] [IsSuccArchimedean α]
+     [DecidableEq α] [@DecidableRel α (· ≤ ·)] [@DecidableRel α (· < ·)] [IsDirected α (· ≥ ·)] :
+     LinearOrder α where
+  le_total a b :=
+    have ⟨c, ha, hb⟩ := directed_of (· ≥ ·) a b
+    le_total_of_ge ha hb
+  decidableEq := inferInstance
+  decidableLE := inferInstance
+  decidableLT := inferInstance
```
Comment (YaelDillies) [inline Mathlib/Order/SuccPred/Archimedean.lean:140]: What happens if
```suggestion
```

**Review by YaelDillies — COMMENTED** [review]
Would you mind moving the new lemmas to a later PR? Can be #16272 if you want

File: Mathlib/Order/SuccPred/Archimedean.lean (hunk 4 above)
Comment (Command-Master) [reply Mathlib/Order/SuccPred/Archimedean.lean]: `fields missing: 'decidableLE'`. You can only give it, but then the derived `decidableEq` and `decidableLT` aren't defeq to the given ones.

**Comment by Command-Master** [comment]
> Would you mind moving the new lemmas to a later PR? Can be #16272 if you want

Alright

**Comment by YaelDillies** [comment]
Thanks!

maintainer merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**